clippy_utils/src/ast/spanless.rs

```diff
@@ -0,0 +1,1768 @@
+        $cx.unordered_eq($lname, $rname)
+    };
+    (@unordered_pat_fields $cx:ident $lname:ident $rname:ident) => {
+        $cx.eq_pat_fields($lname, $rname)
+    };
+}
+macro_rules! hash_enum_field {
+    ($cx:ident _) => {};
+    ($cx:ident $name:ident) => {
+        AstNode::hash($cx, $name)
+    };
+    (@unordered $cx:ident $name:ident) => {
+        $cx.write_unordered_hash($name)
+    };
+    (@unordered_pat_fields $cx:ident $name:ident) => {
+        $cx.write_unordered_hash($name)
+    };
+}
+macro_rules! on_unequal {
+    (@unequal $($val:tt)*) => { $($val)? };
+}
+
+/// Impls for structs which will be implemented matching the variant and comparing
+/// each of it's fields.
+///
+/// The basic form of the macro matches/hashes the variant, and calls `eq` and `hash`
+/// for each of it's fields. Each field must be given a pair of unique names. e.g.
+/// ```ignore
+/// SomeType {
+///     UnitVariant {},
+///     StructVariant {
+///         field1: (l0, r0),
+///         field2: (l1, r1),
+///     },
+///     TupleVariant {
+///         0: (l0, r0),
+///         1: (l1, r1),
+///     },
+/// }
+/// ```
+///
+/// A variant can always be compared as unequal by adding the `#[unequal]`. e.g.
+/// ```ignore
+/// SomeType {
+///     UnitVariant {},
+///     #[unequal] Dummy {},
+/// }
+/// ```
+///
+/// Fields can be ignored by using `_` for both names after the field. e.g.
+/// ```ignore
+/// SomeType {
+///     StructVariant {
+///         field1: (l0, r0),
+///         ignored: (_, _),
+///     },
+/// }
+/// ```
+///
+/// Fields which deref into a slice can be compared in an unordered manner by
+/// adding `#[unordered]`. e.g.
+/// ```ignore
+/// SomeType {
+///     Var {
+///         #[unordered] field: (l0, r0),
+///     },
+/// }
+/// ```
+///
+/// Generics can be used by adding them after the struct name. Constraints must
+/// be parenthesised. e.g.
+/// ```ignore
+/// StructName<T: (Trait1 + Trait2)> {
+///     Var {
+///         field: (l0, r0),
+///     },
+/// }
+/// ```
+macro_rules! impl_enum {
+    ($($name:ident $(<$($gargs:ident $(: ($($gbounds:tt)*))?),*>)? {$(
+        $(#[$unequal:tt])?
+        $vname:ident {
+            $($(#[$unordered:tt])? .$fname:tt: ($lname:tt, $rname:tt),)*
+        },
+    )*})*) => {$(
+        #[allow(non_snake_case, unreachable_code)]
+        impl  $(<$($gargs $(: $($gbounds)*)?),*>)? AstNode for $name $(<$($gargs),*>)? {
+            fn eq(cx: &mut EqCtxt, l: &Self, r: &Self) -> bool {
+                match (l, r) {
+                    $(
+                        (
+                            $name::$vname { $($fname: $lname),*},
+                            $name::$vname { $($fname: $rname),*}
+                        ) => {
+                            $(on_unequal!(@$unequal return false);)?
+                            $(eq_enum_field!($(@$unordered)? cx $lname $rname) &&)* true
+                        },
+                    )*
```

Comment:
Could we use `${concat}` here to allow a similar syntax to the struct definitions?

Reply:
`concat` doesn't work in repetitions and pushing the `concat` into an inner macro means the generated identifier can only the inner macro can access the name due to hygiene rules. Basically neither of these work:

```
// with repetitions: ice
(
  $var { $($name: ${concat(l, $name)}),*  },
  $var { $($name: ${concat(l, $name)}),*  },
) => $(eq!(${concat(l, $name)} ${concat(r, $name)}))&&*,

// with inner: binding names can't be used outside the `name` macro
(
  $var { $($name: name!(l, $name)),*  },
  $var { $($name: name!(l, $name)),*  },
) => $(eq!($name))&&*,
```